### ISSUE TITLE
Add OMPLoopTrans parent class (subclass of ParallelLoopTrans)

### ISSUE THREAD

**Issue by LonelyCat124** [stfc/PSyclone#1431]
The OpenMP loop transformations supported in PSyclone (`OMPTaskloopTrans` and `OMPLoopTrans`) should both check that the loop is of the canonical loop form (https://www.openmp.org/spec-html/5.0/openmpsu40.html#x63-1260002.9.1)

For Fortran style loops this involves (written by Sergi in #1415 ):

- Only BinaryOperation.[ADD, SUB, MUL] + Literals + References are valid in in the incr
- Loop invariant with respect to the outermost associated loop. As discussed the other day I think for a single loop in Fortran, the incr is loop invariant by definition, but I don't think this is true for nested loops then and could be checked in the validate.

An additional check (which we can't always do if the loop increment is not a constant, at which point it isn't PSyclone's issue I think) is that the loop increment is not 0.

I think this should sit as a different class (and not part of `OMPLoopTrans`) as `OMPTaskloopTrans` also uses this, but I don't believe they should be subclasses - perhaps something like `OMPCanonicalLoopTrans`.

One other slight thing about the name `OMPLoopTrans` in general - OpenMP has a loop construct (https://www.openmp.org/spec-html/5.0/openmpsu44.html) - I think its unlikely PSyclone would support this construct (as I'm currently unclear what the use-case of it is), the docstring for OMPLoopTrans should probably also be adjusted from "Adds an OpenMP directive to a loop" to something like "...OpenMP do (or for) directive to a loop"

**Comment by sergisiso** [comment]
>  name `OMPLoopTrans` in general - OpenMP has a loop construct

Oh yeah, that was a bad pick, if fact I add support for this pragma in my las PR and the this transformation name is already used for adding OMPDo/For directives.. We need a better name for this common validation class.

**Comment by LonelyCat124** [comment]
Can we just change to OMPDo/For directives class as well? OR will that break a bunch of code? Should we deprecate it or something and change it in a later PSyclone release?

**Comment by LonelyCat124** [comment]
I started having a look at this a bit, is there an easy way to check if a node has one of many structures? E.>
```
var-outer
  | var-outer + a2
  | a2 + var-outer
  | var-outer - a2
  | a2 - var-outer
  | a1 * var-outer
  | a1 * var-outer + a2
  | a2 + a1 * var-outer
  | a1 * var-outer - a2
  | a2 - a1 * var-outer
  | var-outer * a1
  | var-outer * a1 + a2
  | a2 + var-outer * a1
  | var-outer * a1 - a2
  | a2 - var-outer * a1
```

I assume not and I just have to test each separately but if so it would be helpful.

**Comment by hiker** [comment]
I only did a quick test, but maybe it could be done with sympy? Do I understand correctly that you want to check if the expressions is linear in `var-outer`? Checking if the 2nd derivate is 0 seems to work:
```
>>> sp.diff(sp.diff(a2-a1*var, var), var)
0
>>> sp.diff(sp.diff(a2*var-a1*var, var), var)
0
>>> sp.diff(sp.diff(a2*var-a1*var*var, var), var)
-2*a1
```
Or (thanks to google) there is a degree function in sympy
```
>>> sp.degree(a2*var-a1*var*var, var)
2
>>> sp.degree(a2*var-a1*var, var)
1
>>> sp.degree(a2-a1*var, var)
1
```

Either could be used to implement a `is_linear` function in SymbolicMaths.

**Comment by LonelyCat124** [comment]
Not exactly, I want to check if the expressions conform to exactly one of the allowed formats specified in the OpenMP manual, so as far as I can tell, `a2*var - a1*var` is forbidden despite being 0 derivate (also requires a2 and a1 to be loop-invariate, which i check seperately)

**Comment by rupertford** [comment]
Could you turn the expressions into the above form, then you could reduce the number of tests by canonicalising with sympy,  for example `a1+var-outer == var-outer+a1` and/or perhaps use regular expressions to check for conformance by writing the expression as a string. You should be able to get away with a small number e.g. something a little more constrained than this ... `[+-]?[a1*]?"(var-outer)"[[+-]a2]?`

**Comment by LonelyCat124** [comment]
I could try - it seems to me like all the above variants only need 3 if statements anyway (as it seems PSyIR always has the inner binary operation have the higher priority

**Comment by rupertford** [comment]
Whatever is easiest :-)

**Comment by LonelyCat124** [comment]
Not sure what the cause is, but I'm having some unexpected tests failing now - i think its `dynamo0p3_multigrid_test.py:test_restrict_prolong_chain_anyd`, and it fails when calling `VariablesAccessInfo`, not sure what is happening but I'm crashing at `colourmap` in `DynKern`:
```
>           cmap = invoke.meshes.intergrid_kernels[id(self)].\
                colourmap_symbol.name
E           AttributeError: 'NoneType' object has no attribute 'name'
```

**Comment by hiker** [comment]
> I could try - it seems to me like all the above variants only need 3 if statements anyway (as it seems PSyIR always has the inner binary operation have the higher priority

I don't believe this is necessary true: iirc, if the inner binary operation has a lower priority, the writer will add parenthesis around it automatically, e.g. in fortran.py:
```
                    if precedence(fort_oper) < precedence(parent_fort_oper):
                        # We need brackets to enforce precedence
                        return f"({lhs} {fort_oper} {rhs})"
```
